Bio/Align/exonerate.py

```diff
@@ -0,0 +1,650 @@
+            query_id = "query"
+        try:
+            target_id = target.id
+        except AttributeError:
+            target_id = "target"
+        try:
+            target_molecule_type = target.annotations["molecule_type"]
+        except (AttributeError, KeyError):
+            target_molecule_type = None
+        if target_molecule_type == "protein":
+            target_strand = "."
+        elif target_start <= target_end:
+            target_strand = "+"
+        elif target_start > target_end:
+            target_strand = "-"
+            steps[0, :] = -steps[0, :]
+        try:
+            query_molecule_type = query.annotations["molecule_type"]
+        except (AttributeError, KeyError):
+            query_molecule_type = None
+        if query_molecule_type == "protein":
+            query_strand = "."
+        elif query_start <= query_end:
+            query_strand = "+"
+        elif query_start > query_end:
+            query_strand = "-"
+            steps[1, :] = -steps[1, :]
+        score = alignment.score
+        words = [
+            "cigar:",
+            query_id,
+            str(query_start),
+            str(query_end),
+            query_strand,
+            target_id,
+            str(target_start),
+            str(target_end),
+            target_strand,
+            str(score),
+        ]
+        try:
+            operations = alignment.operations
+        except AttributeError:
+            for step in steps.transpose():
+                target_step, query_step = step
+                if target_step == query_step:
+                    operation = "M"
+                    step = target_step
+                elif query_step == 0:
+                    operation = "D"  # Deletion
+                    step = target_step
+                elif target_step == 0:
+                    operation = "I"  # Insertion
+                    step = query_step
+                elif (
+                    target_molecule_type != "protein"
+                    and query_molecule_type == "protein"
+                ):
+                    operation = "M"
+                    step = target_step
+                elif (
+                    target_molecule_type == "protein"
+                    and query_molecule_type != "protein"
+                ):
+                    operation = "M"
+                    step = query_step
+                else:
+                    raise ValueError(
+                        "Unexpected step target %d, query %d for molecule type %s, %s"
+                        % (
+                            target_step,
+                            query_step,
+                            target_molecule_type,
+                            query_molecule_type,
+                        )
+                    )
+                words.append(operation)
+                words.append(str(step))
+        else:
+            for step, operation in zip(steps.transpose(), operations.decode()):
```

Comment:
Is it correctly understood that we go into this `else` block if the incoming alignment objects was created by a non-Exonerate parser (therefore having no `operations` attribute)?

Reply:
That is correct. To save memory, when parsing Exonerate files we could also store the `operations` attribute if it contains any non-trivial information. I'll leave that for a future update.